pkg/operator2/configobservation/oauth/observe_idps_test.go

```diff
@@ -0,0 +1,193 @@
+package oauth
+
+import (
+	"fmt"
+	"testing"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/diff"
+	"k8s.io/client-go/tools/cache"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configlistersv1 "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/openshift/cluster-authentication-operator/pkg/operator2/configobservation"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resourcesynccontroller"
+)
+
+type mockResourceSyncer struct {
+	t      *testing.T
+	synced map[string]string
+}
+
+func (rs *mockResourceSyncer) SyncConfigMap(destination, source resourcesynccontroller.ResourceLocation) error {
```

Comment:
(No action required) Would it be make sense to avoid mocking by separating api calls from logic, e.g.

```golang

type idpResult {
  currentSyncData  *datasync.ConfigSyncData
  observedSyncData *datasync.ConfigSyncData
  observedConfig   map[string]interface{}
  eventMessage     string
}

// With dependencies on the API, this function is ideally validated in e2e
func ObserveIdentityProviders(genericlisters configobserver.Listers, recorder events.Recorder, previouslyObservedConfig map[string]interface{}) (map[string]interface{}, []error) {
    listers := genericlisters.(configobservation.Listers)
    resourceSyncer := genericlisters.ResourceSyncer()

    oauthConfig, _ := listers.OAuthLister.Get("cluster")
    ... 

    result, _ := computeIdpResult(oathConfig, previouslyObservedConfig)
    ...

    datasync.HandleIdPConfigSync(resourceSyncer, result.currentSyncData, result.observedSyncData)
    
    if len(result.eventMessage) {
        recorder.Eventf("ObserveIdentityProviders", result.eventMessage)
    }

    return result.observedConfig, nil
}

// With no dependencies on the API, the logic of this function is ideally validated with unit testing
func computeIdpResult(oathConfig *v1.OAuth, previousConfig map[string]interface{}) (*idpResult, error) {
  ...
}
```

Reply:
Both seem to bear more or less similar results I guess, I find the mocking to be quite elegant way to test this, but I can see how your approach separates the logical unit outside so that it is easier to raise expectations over it.